### PR TITLE
Add gaze detection

### PR DESCRIPTION
**Pull request by PacificDou**
# Description

Add gaze detection to core models of Roboflow.
It actually has two models: [mediapipe](https://developers.google.com/mediapipe/solutions/vision/face_detector) for face detection from Google, [L2CS](https://github.com/Ahmednull/L2CS-Net) for gaze detection.
The trained models will need to be downloaded via Roboflow API (pull request already sent).

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

This has been tested using image url and local images (base64 encoded).
A simple test case has been included in test/gaze_test.py

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: inference/core/data_models.py [hunk 1]
```diff
@@ -699,3 +713,92 @@ class SamSegmentationResponse(BaseModel):
     time: float = Field(
         description="The time in seconds it took to produce the segmentation including preprocessing"
     )
+
+
+class FaceDetectionPrediction(BaseModel):
```
Comment (paulguerrie) [inline inference/core/data_models.py:718]: Seems like `FaceDetectionPrediction` could inherit from `ObjectDetectionPrediction`

File: inference/core/data_models.py [hunk 2]
```diff
@@ -699,3 +713,92 @@ class SamSegmentationResponse(BaseModel):
...
+    """
+
+    api_key: Optional[str] = ApiKey
+    gaze_version_id: Optional[str] = Field(
+        default=GAZE_VERSION_ID,
+        example="l2cs",
+        description="The version ID of Gaze to be used for this request. Must be one of l2cs.",
+    )
+
+    is_cropped_face_image: Optional[bool] = Field(
```
[the remark this answers is not in the document]
Comment (paulguerrie) [reply inference/core/data_models.py:779]: And default it to `True`

File: inference/core/data_models.py [hunk 3]
```diff
@@ -699,3 +713,92 @@ class SamSegmentationResponse(BaseModel):
...
+        predictions (List[GazeDetectionPrediction]): List of gaze detection predictions.
+        time_total (float): The total processing time (second).
+        time_load_img (float): The processing time for loading image (second).
+        time_face_det (float): The processing time for face detection (second).
+        time_gaze_det (float): The processing time for gaze detection (second).
+    """
+
+    predictions: List[GazeDetectionPrediction]
+
+    time_total: float = Field(description="The total processing time (second)")
```
Comment (paulguerrie) [inline inference/core/data_models.py:801]: Change `total_time` to `time` to better match other response models

File: inference/models/gaze/gaze.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+        if REQUIRED_ONNX_PROVIDERS:
+            available_providers = onnxruntime.get_available_providers()
+            for provider in REQUIRED_ONNX_PROVIDERS:
+                if provider not in available_providers:
+                    raise OnnxProviderNotAvailable(
+                        f"Required ONNX Execution Provider {provider} is not availble. Check that you are using the correct docker image on a supported device."
+                    )
+
+        # init face detector
+        self.face_detector = mp.tasks.vision.FaceDetector.create_from_options(
```
Comment (paulguerrie) [inline inference/models/gaze/gaze.py:123]: Is it possible to go ahead and get the FaceDetector into onnx format? If so, let's do that. If not, let's just leave everything as pytorch. One of the main advantage of using onnx is to not require pytorch as a dependancy. So being half in pytorch, half in onnx isn't buying us much.

File: inference/models/gaze/gaze.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+from inference.core.env import (
+    GAZE_MAX_BATCH_SIZE,
+    REQUIRED_ONNX_PROVIDERS,
+    TENSORRT_CACHE_PATH,
+)
+from inference.core.exceptions import OnnxProviderNotAvailable
+from inference.core.models.roboflow import OnnxRoboflowCoreModel
+
+
+class L2C2Wrapper(L2CS):
```
[the remark this answers is not in the document]
Comment (paulguerrie) [reply inference/models/gaze/gaze.py:36]: Oops! Never mind! I see you already have done this!

File: inference/models/gaze/gaze.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+            imgH (int): The height (px) of original image.
+            time_load_img (float): The processing time for loading image.
+            time_face_det (float): The processing time for face detection.
+            time_gaze_det (float): The processing time for gaze detection.
+
+        Returns:
+            GazeDetectionInferenceResponse: The response object including the detected faces and gazes info.
+        """
+        predictions = []
+        for i in range(len(faces)):
```
Comment (paulguerrie) [inline inference/models/gaze/gaze.py:185]: Could be `for face, gaze in zip(faces,gazes)`

File: inference/models/gaze/gaze.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        """Detect faces and gazes in an image.
+
+        Args:
+            pil_imgs (List[Image.Image]): The PIL image list, each image is a cropped facial image.
+
+        Returns:
+            List[Tuple[float, float]]: Pitch (degree) and Yaw (degree).
+        """
+        ret = []
+        for i in range(0, len(pil_imgs), GAZE_MAX_BATCH_SIZE):
```
Comment (paulguerrie) [inline inference/models/gaze/gaze.py:221]: Nice! Great way of respecting the GAZE_MAX_BATCH_SIZE without dropping any predictions.

**Comment by paulguerrie** [comment]
Can you add an entry here: https://github.com/roboflow/inference/blob/0ae0fe9d9e8ae2e78442ae1d9a337699317a16b0/.release/pypi/inference.core.setup.py#L43

That points to the new requirements.gaze.txt file?

File: inference/models/gaze/gaze.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+        t1 = perf_counter()
+        faces = []
+        for pil_img in pil_imgs:
+            if not request.is_cropped_face_image:
+                mp_img = mp.Image(image_format=mp.ImageFormat.SRGB, data=np.asarray(pil_img))
+                faces_per_img = self.face_detector.detect(mp_img).detections
+            else:
+                faces_per_img = [
+                    Detection(bounding_box=BoundingBox(origin_x=0, origin_y=0, width=pil_img[0], height=pil_img[1]),
+                              categories=[Category(score=1.0, category_name="face")], keypoints=[])]
```
Comment (bigbitbus) [inline inference/models/gaze/gaze.py:293]: Is the "score=1.0" then threshold for a face detection? Does it need to be a tunable parameter or is it binary with 1.0  a positive and everything else a negative detection?

**Review by bigbitbus — COMMENTED** [review]
Nice work! Had a couple of questions in the comments.

File: inference/models/gaze/gaze.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+        # init face detector
+        self.face_detector = mp.tasks.vision.FaceDetector.create_from_options(
+            mp.tasks.vision.FaceDetectorOptions(
+                base_options=mp.tasks.BaseOptions(model_asset_path=self.cache_file("mediapipe_face_detector.tflite")),
+                running_mode=mp.tasks.vision.RunningMode.IMAGE
+            )
+        )
+
+        # additional settings for gaze detection
+        self._gaze_transformations = transforms.Compose([
```
[the remark this answers is not in the document]
Comment (bigbitbus) [reply inference/models/gaze/gaze.py:131]: Looks good after discussion.

File: inference/models/gaze/gaze.py [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+        # TODO: face detection for batch
+        t1 = perf_counter()
+        faces = []
+        for pil_img in pil_imgs:
+            if not request.is_cropped_face_image:
+                mp_img = mp.Image(image_format=mp.ImageFormat.SRGB, data=np.asarray(pil_img))
+                faces_per_img = self.face_detector.detect(mp_img).detections
+            else:
+                faces_per_img = [
+                    Detection(bounding_box=BoundingBox(origin_x=0, origin_y=0, width=pil_img[0], height=pil_img[1]),
```
Comment (bigbitbus) [inline inference/models/gaze/gaze.py:292]: Check if this is the center or not.

**Review by bigbitbus — APPROVED** [review]

File: inference/core/data_models.py [hunk 11]
```diff
@@ -374,6 +374,20 @@ class Point(BaseModel):
     y: float = Field(description="The y-axis pixel coordinate of the point")
 
 
+class Point3D(BaseModel):
```
[the remark this answers is not in the document]
Comment (PacificDou) [reply inference/core/data_models.py:377]: sure, will update it accordingly

File: inference/models/gaze/gaze.py (hunk 10 above)
Comment (PacificDou) [reply inference/models/gaze/gaze.py:292]: The Detection object defined in Mediapipe is using top left corner. However, in our model, we'll return the center point of bounding box.

File: inference/models/gaze/gaze.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+        pitch_degree = pitch_predicted * np.pi / 180.0
+
+        yaw_predicted = self._gaze_softmax(gaze_yaw)
+        yaw_predicted = torch.sum(yaw_predicted * idx_tensor, dim=1) * 4 - 180
+        yaw_degree = yaw_predicted * np.pi / 180.0
+
+        return pitch_degree, yaw_degree
+
+
+class GazeOnnxRoboflowCoreModel(OnnxRoboflowCoreModel):
```
[the remark this answers is not in the document]
Comment (PacificDou) [reply inference/models/gaze/gaze.py:79]: I'll arrange the classes in alphabetical order. For methods in class, I think it's better to keep those close if they are being used for same/similar reasons?
For example, L2C2Wrapper.forward is for inference, while L2C2Wrapper.load_L2CS_model and L2C2Wrapper.saveas_ONNX_model are for converting models to ONNX format.

In other words, I think it's might be better to group methods by functionality first, then sort them according to alphabetical order (or invoking order??), what do you think?

File: inference/core/data_models.py (hunk 3 above)
Comment (PacificDou) [reply inference/core/data_models.py:801]: How about we use the following definition:
time: Union[float, List[float]]
a. in production, we only return a single float for the processing time;
b. when necessary in debug, we can return a list of processing times for each crucial step

File: inference/models/gaze/gaze.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (paulguerrie) [reply inference/models/gaze/gaze.py:79]: I did some searching to see if there was a standard convention and there does not appear to be one. The rest of the module lists methods in alphabetical order so maybe we should follow suite to match the other models in `inference`? Also, I think the models are granular enough that the the model class itself can be thought of a convenient logical grouping. Especially when working on code that I was not the original author of, I find it convenient when I can rely on alphabetical order to navigate a file.

File: inference/core/data_models.py (hunk 3 above)
Comment (paulguerrie) [reply inference/core/data_models.py:801]: Let's keep `time` just a `float` to match other model responses. You could make the other time fields optional and only return them in debug mode if you want to keep them around. But we don't currently have any sort of debug mode options, so might be better to just remove them for now.

File: inference/models/gaze/gaze.py (hunk 4 above)
Comment (PacificDou) [reply inference/models/gaze/gaze.py:123]: The model itself can be converted to onnx model by using [tf2onnx](https://github.com/onnx/tensorflow-onnx). 
However, in mediapipe, they load the tflite model as an internal object, and hide image preprocessing and postprocessing.
So, even we can convert the model, without knowing how the whole pipeline is working, it's almost impossible to use onnx instead of mediepipe for those models. 
I'd prefer to use mediapipe for now, and let's see whether we can unify everything later.

File: inference/models/gaze/gaze.py [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+            score = face.categories[0].score
+
+            prediction = GazeDetectionPrediction(
+                face=FaceDetectionPrediction(x=x_center, y=y_center, width=bbox.width, height=bbox.height,
+                                             confidence=score, class_name="face", landmarks=landmarks),
+                pitch=gaze[0], yaw=gaze[1])
+            predictions.append(prediction)
+
+        response = GazeDetectionInferenceResponse(predictions=predictions,
+                                                  time_total=(time_load_img + time_face_det + time_gaze_det),
```
[the remark this answers is not in the document]
Comment (PacificDou) [reply inference/models/gaze/gaze.py:206]: Will keep the time field. For additional debug time info, will set them as optional.

**Review by bigbitbus — COMMENTED** [review]
Already reviewed and approved this request, but will let @paulguerrie and @PacificDou work out the changes before formal approval.

File: inference/models/gaze/gaze.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+        Args:
+            request (GazeDetectionInferenceRequest): The request object containing the image.
+
+        Returns:
+            List[GazeDetectionInferenceResponse]: The list of response objects containing the faces and corresponding gazes.
+        """
+        if isinstance(request.image, list):
+            if len(request.image) > GAZE_MAX_BATCH_SIZE:
+                raise ValueError(
+                    f"The maximum number of images that can be embedded at once is {GAZE_MAX_BATCH_SIZE}"
```
Comment (paulguerrie) [inline inference/models/gaze/gaze.py:203]: Update message to "The maximum number of images that can be inferred with gaze detection at one time is {GAZE_MAX_BATCH_SIZE}"

**Review by paulguerrie — APPROVED** [review]
Looks great! Great PR @PacificDou!

**Comment by paulguerrie** [comment]
@PacificDou , looks like you need to run `make style` in the repo, then commit the changes. (likely import order updates)

**Review by bigbitbus — APPROVED** [review]
Tested together, lgtm. Super ship @PacificDou
